### PR TITLE
add new env config options for OTLP exporter

### PR DESCRIPTION
**Pull request by seankhliao**
spec: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options

Add the `INSECURE` `CLIENT_CERTIFICATE` and `CLIENT_KEY` env options for the otlp exporters

### CODE REVIEW

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 1]
```diff
@@ -59,6 +59,17 @@ func WithString(n string, fn func(string)) func(e *EnvOptionsReader) {
 	}
 }
 
+// WithBool retrieves the specified config and passes it to ConfigFn as a boolean.
+func WithBool(n string, fn func(bool)) func(e *EnvOptionsReader) {
+	return func(e *EnvOptionsReader) {
+		if v, ok := e.GetEnvValue(n); ok {
+			if b, err := strconv.ParseBool(v); err == nil {
```
Comment (Aneurysm9) [inline exporters/otlp/internal/envconfig/envconfig.go:66]: Not sure we can use this.  [Boolean env values](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#boolean-value) MUST treat case-insensitive "true" as `true` and everything else as `false`.

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 2]
```diff
@@ -91,18 +102,38 @@ func WithURL(n string, fn func(*url.URL)) func(e *EnvOptionsReader) {
 }
 
 // WithTLSConfig retrieves the specified config and passes it to ConfigFn as a crypto/tls.Config.
```
Comment (Aneurysm9) [inline exporters/otlp/internal/envconfig/envconfig.go:104]: Be sure to update the godoc if you're going to change this function name.

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 3]
```diff
@@ -91,18 +102,38 @@ func WithURL(n string, fn func(*url.URL)) func(e *EnvOptionsReader) {
...
-				if c, err := createTLSConfig(b); err == nil {
+				if c, err := createCertPool(b); err == nil {
 					fn(c)
 				}
 			}
 		}
 	}
 }
 
+func WithClientCert(nc, nk string, fn func(tls.Certificate)) func(e *EnvOptionsReader) {
```
Comment (Aneurysm9) [inline exporters/otlp/internal/envconfig/envconfig.go:117]: Please add some godoc explaining this function and its parameters.

File: exporters/otlp/internal/envconfig/envconfig_test.go [hunk 4]
```diff
@@ -134,6 +144,52 @@ func TestEnvConfig(t *testing.T) {
...
+					}
+					return ""
+				},
+			},
+			configs: []ConfigFn{
+				WithBool("HELLO", func(b bool) {
+					options = append(options, testOption{TestBool: b})
+				}),
+			},
+			expectedOptions: []testOption{},
```
Comment (Aneurysm9) [inline exporters/otlp/internal/envconfig/envconfig_test.go:191]: This should have `{TestBool: false}` as a result.

File: exporters/otlp/internal/envconfig/envconfig.go (hunk 1 above)
Comment (seankhliao) [reply exporters/otlp/internal/envconfig/envconfig.go]: used a ToLower comparison with `"true"`

File: exporters/otlp/internal/envconfig/envconfig.go (hunk 2 above)
Comment (seankhliao) [reply exporters/otlp/internal/envconfig/envconfig.go]: done

File: exporters/otlp/internal/envconfig/envconfig_test.go (hunk 4 above)
Comment (seankhliao) [reply exporters/otlp/internal/envconfig/envconfig_test.go]: done

**Comment by seankhliao** [comment]
not sure what to do with coverage as the other places i modified don't seem to have tests in place

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 5]
```diff
@@ -59,6 +59,16 @@ func WithString(n string, fn func(string)) func(e *EnvOptionsReader) {
 	}
 }
 
+// WithBool retrieves the specified config and passes it to ConfigFn as a boolean.
```
Comment (MrAlias) [inline exporters/otlp/internal/envconfig/envconfig.go:62]:
```suggestion
// WithBool returns a ConfigFn that reads the environment variable n and if it exists passes its parsed bool value to fn.
```

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 6]
```diff
@@ -90,19 +100,42 @@ func WithURL(n string, fn func(*url.URL)) func(e *EnvOptionsReader) {
...
+func WithClientCert(nc, nk string, fn func(tls.Certificate)) func(e *EnvOptionsReader) {
+	return func(e *EnvOptionsReader) {
+		vc, okc := e.GetEnvValue(nc)
+		vk, okk := e.GetEnvValue(nk)
+		if !okc || !okk {
+			return
+		}
+		cert, errc := e.ReadFile(vc)
+		key, errk := e.ReadFile(vk)
+		if errc != nil && errk != nil {
```
Comment (MrAlias) [inline exporters/otlp/internal/envconfig/envconfig.go:128]: We should likely log these errors so the user known why their connection doesn't have TLS setup correctly: 

https://github.com/open-telemetry/opentelemetry-go/blob/e8023fab22dc1cf95b47dafcc8ac8110c6e72da1/internal/global/internal_logging.go#L52

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 7]
```diff
@@ -90,19 +100,42 @@ func WithURL(n string, fn func(*url.URL)) func(e *EnvOptionsReader) {
...
+			return
+		}
+		crt, err := tls.X509KeyPair(cert, key)
+		if err != nil {
```
Comment (MrAlias) [inline exporters/otlp/internal/envconfig/envconfig.go:132]: Similar here regarding logging.

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 8]
```diff
@@ -59,6 +59,16 @@ func WithString(n string, fn func(string)) func(e *EnvOptionsReader) {
 	}
 }
 
+// WithBool retrieves the specified config and passes it to ConfigFn as a boolean.
+func WithBool(n string, fn func(bool)) func(e *EnvOptionsReader) {
```
Comment (MrAlias) [inline exporters/otlp/internal/envconfig/envconfig.go:63]:
```suggestion
func WithBool(n string, fn func(bool)) ConfigFn {
```

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 9]
```diff
@@ -90,19 +100,42 @@ func WithURL(n string, fn func(*url.URL)) func(e *EnvOptionsReader) {
 	}
 }
 
-// WithTLSConfig retrieves the specified config and passes it to ConfigFn as a crypto/tls.Config.
-func WithTLSConfig(n string, fn func(*tls.Config)) func(e *EnvOptionsReader) {
+// WithCertPool retrieves the specified config and passes it to ConfigFn as a crypto/x509.CertPool
+// constructed from reading the certificate at the given path.
+func WithCertPool(n string, fn func(*x509.CertPool)) func(e *EnvOptionsReader) {
```
Comment (MrAlias) [inline exporters/otlp/internal/envconfig/envconfig.go:105]:
```suggestion
func WithCertPool(n string, fn func(*x509.CertPool)) ConfigFn {
```

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 10]
```diff
@@ -90,19 +100,42 @@ func WithURL(n string, fn func(*url.URL)) func(e *EnvOptionsReader) {
...
 					fn(c)
 				}
 			}
 		}
 	}
 }
 
+// WithCertPool retrieves the specified configs and passes it to ConfigFn as a crypto/tls.Certificate
+// constructed from reading the key pair at the given paths. Both files must exist.
+func WithClientCert(nc, nk string, fn func(tls.Certificate)) func(e *EnvOptionsReader) {
```
Comment (MrAlias) [inline exporters/otlp/internal/envconfig/envconfig.go:119]:
```suggestion
func WithClientCert(nc, nk string, fn func(tls.Certificate)) ConfigFn {
```

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 11]
```diff
@@ -90,19 +100,42 @@ func WithURL(n string, fn func(*url.URL)) func(e *EnvOptionsReader) {
...
+				if c, err := createCertPool(b); err == nil {
 					fn(c)
 				}
 			}
 		}
 	}
 }
 
+// WithCertPool retrieves the specified configs and passes it to ConfigFn as a crypto/tls.Certificate
+// constructed from reading the key pair at the given paths. Both files must exist.
```
Comment (MrAlias) [inline exporters/otlp/internal/envconfig/envconfig.go:118]:
```suggestion
// WithClientCert returns a ConfigFn that reads the environment variable nc and nk as filepaths to a client certificate and key pair. If they exists, they are parsed as a crypto/tls.Certificate and it is passed to fn.
```

File: exporters/otlp/internal/envconfig/envconfig.go [hunk 12]
```diff
@@ -90,19 +100,42 @@ func WithURL(n string, fn func(*url.URL)) func(e *EnvOptionsReader) {
 	}
 }
 
-// WithTLSConfig retrieves the specified config and passes it to ConfigFn as a crypto/tls.Config.
-func WithTLSConfig(n string, fn func(*tls.Config)) func(e *EnvOptionsReader) {
+// WithCertPool retrieves the specified config and passes it to ConfigFn as a crypto/x509.CertPool
+// constructed from reading the certificate at the given path.
```
Comment (MrAlias) [inline exporters/otlp/internal/envconfig/envconfig.go:104]:
```suggestion
// WithCertPool returns a ConfigFn that reads the environment variable n as a filepath to a TLS certificate pool. If it exists, it is parsed as a crypto/x509.CertPool and it is passed to fn.
```

File: exporters/otlp/internal/envconfig/envconfig.go (hunk 6 above)
Comment (seankhliao) [reply exporters/otlp/internal/envconfig/envconfig.go]: Done, also added logging for other silently ignored errors in the file

**Review by MrAlias — APPROVED** [review]
:1st_place_medal:

**Review by hanyuancheung — APPROVED** [review]

**Comment by hanyuancheung** [comment]
Seems the changelog has conflicts, please resolve them.

**Comment by seankhliao** [comment]
resolved

**Comment by MrAlias** [comment]
@Aneurysm9 does this look good to you?

**Review by Aneurysm9 — APPROVED** [review]

**Pull request merged**